### DISCUSSION TITLE
Troubleshooting: [setAuthCookies] Failed to verify the ID token. Cannot authenticate the user or get a refresh token.

### DISCUSSION THREAD

**Discussion by jimjeffers** [gladly-team/next-firebase-auth#569] [Q&A]
I'm successfully authenticating on the client side but my server side session repeatedly throws the same error:

```
next-firebase-auth: [setAuthCookies] Attempting to set auth cookies.
next-firebase-auth: [setAuthCookies] Failed to verify the ID token. Cannot authenticate the user or get a refresh token.
next-firebase-auth: [setAuthCookies] Set auth cookies. The user is not authenticated.
```

I logged my request headers on the `/api/login` route to ensure the proper credentials are getting sent over and everything looks well:

```js
{
  host: 'localhost:3000',
  origin: 'http://localhost:3000',
  'accept-encoding': 'gzip, deflate',
  cookie: 'edna-explorer.AuthUser=IntcImlkXCI6bnVsbCxcImNsYWltc1wiOnt9LFwiZW1haWxcIjpudWxsLFwiZW1haWxWZXJpZmllZFwiOmZhbHNlLFwicGhvbmVOdW1iZXJcIjpudWxsLFwiZGlzcGxheU5hbWVcIjpudWxsLFwicGhvdG9VUkxcIjpudWxsLFwiY2xpZW50SW5pdGlhbGl6ZWRcIjpmYWxzZX0i; edna-explorer.AuthUser.sig=eDQX8gIjvtvS8J19Jr_Z0VliacY; edna-explorer.AuthUserTokens=IntcImlkVG9rZW5cIjpudWxsLFwicmVmcmVzaFRva2VuXCI6bnVsbH0i; edna-explorer.AuthUserTokens.sig=bIswACcQSbMkZHa868AIPuLfdPQ; ExampleApp.AuthUser=IntcImlkXCI6bnVsbCxcImNsYWltc1wiOnt9LFwiZW1haWxcIjpudWxsLFwiZW1haWxWZXJpZmllZFwiOmZhbHNlLFwicGhvbmVOdW1iZXJcIjpudWxsLFwiZGlzcGxheU5hbWVcIjpudWxsLFwicGhvdG9VUkxcIjpudWxsLFwiY2xpZW50SW5pdGlhbGl6ZWRcIjpmYWxzZX0i; ExampleApp.AuthUser.sig=ee5q0JaGIipVQlPHGkcBlzhU2Vs; ExampleApp.AuthUserTokens=IntcImlkVG9rZW5cIjpudWxsLFwicmVmcmVzaFRva2VuXCI6bnVsbH0i; ExampleApp.AuthUserTokens.sig=nXcwPwI6iIMP4wUOUWfBo-mQrug',
  'content-length': '0',
  connection: 'keep-alive',
  accept: '*/*',
  'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.0 Safari/605.1.15',
  authorization: 'eyJhbGciOiJSUzI1NiIsImtpZCI6ImRjMzdkNTkzNjVjNjIyOGI4Y2NkYWNhNTM2MGF...',
  referer: 'http://localhost:3000/home',
  'accept-language': 'en-US,en;q=0.9'
}
```

Any suggestions on how to diagnose or quickly make the necessary server side call manually to see why firebase or the auth framework is complaining here?

I'm using the following versions in my package.json:

```json
"firebase": "^9.13.0",
    "firebase-admin": "^11.2.0",
    "next": "^12.3.1",
    "next-absolute-url": "^1.2.2",
    "next-firebase-auth": "^1.0.0-canary.17",
 ```

**Comment by kmjennison** [comment]
@jimjeffers Thanks for the detailed info!

It's interesting that this doesn't have more info in the logs. The `setAuthCookies` call [here](https://github.com/gladly-team/next-firebase-auth/blob/v1.x/src/setAuthCookies.js#L29-L37) is logging the debug error, and many of the common errors in `getCustomIdAndRefreshTokens` [here](https://github.com/gladly-team/next-firebase-auth/blob/3bc813ebaff8851ce6f74abaa658656daa8da463/src/firebaseAdmin.js#L174-L187) are logged. You must be hitting an error that's not handled there.

The most direct way to see what you're running into would be to log the error created [here](https://github.com/gladly-team/next-firebase-auth/blob/v1.x/src/setAuthCookies.js#L33). If possible, could you try cloning this repo, modifying that catch block, and running a local version of this library in your app? Here's an overview of how to do this:
https://github.com/gladly-team/next-firebase-auth/blob/v1.x/CONTRIBUTING.md#developing-with-a-local-version-of-the-package

**Comment by JamesDonnelly** [comment]
I was struggling with the same problem today. It turned out to be my `FIREBASE_PRIVATE_KEY` not being decoded correctly. The canary version of next-firebase-auth gave no information at all, but after downgrading to Firebase 8 my `/api/login` route started returning a 500 instead of the 200 it was before. The exception being thrown was:

> FirebaseAppError: Failed to parse private key: Error: Invalid PEM formatted message.

There are some quick solutions to this in https://github.com/gladly-team/next-firebase-auth/discussions/95.

Hopefully this is the same issue you're having. I'm now back on canary with Firebase 9 and it's all working as expected.

**Reply by demc** [reply]
Thanks for this... sigh...
